gen/builders/tipsets.go

```diff
@@ -0,0 +1,95 @@
+package builders
+
+import (
+	"github.com/ipfs/go-cid"
+
+	"github.com/filecoin-project/test-vectors/schema"
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+)
+
+// TipsetSeq is a sequence of tipsets to be applied during the test.
+// TipsetSeq provides methods to build a sequence of tipsets, such as
+// adding new tipsets and null rounds.
+type TipsetSeq struct {
+	tipsets []*Tipset
+	epoch   abi.ChainEpoch
+
+	// msgIdx is an index that stores unique messages enlisted in blocks in this
+	// tipset sequence.
+	msgIdx map[cid.Cid]*ApplicableMessage
+}
+
+type Tipset struct {
+	tss *TipsetSeq
+
+	// PostStateRoot stores the state root CID after applying this tipset.
+	// It can be used with Asserter#AtState to obtain an asserter against that
+	// state root.
+	PostStateRoot cid.Cid
+
+	schema.Tipset
+}
+
+type Block = schema.Block
+
+// NewTipsetSeq returns a new TipSetSeq object initialized at the provided
+// epoch.
+func NewTipsetSeq(initialEpoch abi.ChainEpoch) *TipsetSeq {
+	return &TipsetSeq{
+		epoch:  initialEpoch,
+		msgIdx: make(map[cid.Cid]*ApplicableMessage),
+	}
+}
+
+// All returns all tipsets that have been registered.
+func (tss *TipsetSeq) All() []*Tipset {
+	return tss.tipsets
+}
+
+// Messages returns all ApplicableMessages that have been included in blocks,
+// in no particular order.
+func (tss *TipsetSeq) Messages() []*ApplicableMessage {
+	msgs := make([]*ApplicableMessage, 0, len(tss.msgIdx))
+	for _, msg := range tss.msgIdx {
+		msgs = append(msgs, msg)
+	}
+	return msgs
+}
+
+// Next enrols a new Tipset, with the supplied base fee, and advances the epoch
+// by 1.
+func (tss *TipsetSeq) Next(baseFee abi.TokenAmount) *Tipset {
+	ts := &Tipset{
+		tss: tss,
+		Tipset: schema.Tipset{
+			Epoch:   tss.epoch,
+			BaseFee: baseFee,
+		},
+	}
+	tss.tipsets = append(tss.tipsets, ts)
+	tss.epoch++ // advance the epoch.
+	return ts
+}
+
+// NullRounds enrols as many null rounds as indicated, advancing the epoch by
+// the count.
+func (tss *TipsetSeq) NullRounds(count uint64) {
+	tss.epoch += abi.ChainEpoch(count)
+}
+
+// Block adds a new block to this tipset, produced by the indicated miner, with
+// the supplied wincount, and containing the listed (and previously staged)
+// msgIdx.
+func (ts *Tipset) Block(miner Miner, winCount int64, msgs ...*ApplicableMessage) {
```

Comment:
It does feel like it might be cleaner to have this method on `TipsetSeq` and remove the back reference on each `Tipset` to the `TipsetSeq` containing it.

Reply:
That would make the API very unflexible. Right now you can create all your tipsets in one batch, and then start adding blocks to them. If you move this method up, you would need to follow the pattern, which is also rather obscure and has hidden assumptions, which I'd like to avoid.

```go
v.Tipsets.Next()
v.Tipsets.Block() <= this is implicitly associated with the previous tipset, but it's not visible
v.Tipsets.Next()
v.Tipsets.Block()
```

Unless you mean:

```go
ts := v.Tipsets.Next()
v.Tipsets.Block(ts, ...)
```

Which I think is a bit messier in terms of DX. Priority here is DX, so I'm inclined to keep as-is. (and the backreference doesn't hurt)